src/joystick/hidapi/SDL_hidapi_sinput.c

```diff
@@ -0,0 +1,745 @@
+{
+    SDL_DriverSInput_Context *ctx = (SDL_DriverSInput_Context *)device->context;
+
+    FeatureFlagsUnpack(device, data);
+
+#if defined(DEBUG_SINPUT_INIT)
+    SDL_Log("Accelerometer Range: %d", ctx->accelRange);
+#endif
+
+
+#if defined(DEBUG_SINPUT_INIT)
+    SDL_Log("Gyro Range: %d", ctx->gyroRange);
+#endif
+
+    ctx->accelScale = CalculateAccelScale(ctx->accelRange);
+    ctx->gyroScale  = CalculateGyroScale(ctx->gyroRange);
+
+    ctx->feature_flags_obtained = true;
+
+    // Set player number, finalizing the setup
+    Uint8 playerLedCommand[SINPUT_DEVICE_REPORT_COMMAND_SIZE] = { SINPUT_DEVICE_REPORT_ID_OUTPUT_CMDDAT, SINPUT_DEVICE_COMMAND_PLAYERLED, ctx->player_idx };
+    int playerNumBytesWritten = SDL_HIDAPI_SendRumble(device, playerLedCommand, SINPUT_DEVICE_REPORT_COMMAND_SIZE);
+
+    if (playerNumBytesWritten < 0) {
+        SDL_SetError("SInput device player led command could not write");
+    }
+}
+
+static bool HIDAPI_DriverSInput_InitDevice(SDL_HIDAPI_Device *device)
+{
+#if defined(DEBUG_SINPUT_INIT)
+    SDL_Log("SInput device Init");
+#endif
+
+    SDL_DriverSInput_Context *ctx = (SDL_DriverSInput_Context *)SDL_calloc(1, sizeof(*ctx));
+    if (!ctx) {
+        return false;
+    }
+
+    ctx->device = device;
+    device->context = ctx;
+
+#if !defined(DEBUG_SINPUT_INIT)
+    ctx->haptics_supported = true;
+    ctx->player_leds_supported = true;
+    ctx->accelerometer_supported = true;
+    ctx->gyroscope_supported = true;
+    ctx->left_analog_stick_supported = true;
+    ctx->right_analog_stick_supported = true;
+    ctx->left_analog_trigger_supported = true;
+    ctx->right_analog_trigger_supported = true;
+    ctx->accelRange = SINPUT_DEFAULT_ACCEL_SENS;
+    ctx->gyroRange = SINPUT_DEFAULT_GYRO_SENS;
+
+    ctx->accelScale = CalculateAccelScale(ctx->accelRange);
+    ctx->gyroScale  = CalculateGyroScale(ctx->gyroRange);
+#endif 
+    
+    return HIDAPI_JoystickConnected(device, NULL);
+}
+
+static int HIDAPI_DriverSInput_GetDevicePlayerIndex(SDL_HIDAPI_Device *device, SDL_JoystickID instance_id)
+{
+    return -1;
+}
+
+static void HIDAPI_DriverSInput_SetDevicePlayerIndex(SDL_HIDAPI_Device *device, SDL_JoystickID instance_id, int player_index)
+{
+    player_index = SDL_clamp(player_index+1, 0, 255);
+    Uint8 player_num = (Uint8)player_index;
+
+    SDL_DriverSInput_Context *ctx = (SDL_DriverSInput_Context *)device->context;
+
+    ctx->player_idx = player_num;
+}
+
+#ifndef DEG2RAD
+#define DEG2RAD(x) ((float)(x) * (float)(SDL_PI_F / 180.f))
+#endif
+
+
+static bool HIDAPI_DriverSInput_OpenJoystick(SDL_HIDAPI_Device *device, SDL_Joystick *joystick)
+{
+#if defined(DEBUG_SINPUT_INIT)
+    SDL_Log("SInput device Open");
+#endif
+
+    SDL_DriverSInput_Context *ctx = (SDL_DriverSInput_Context *)device->context;
+
+    SDL_AssertJoysticksLocked();
+
+    ctx->joystick = joystick;
+
+    SDL_zeroa(ctx->last_state);
+
+    joystick->nbuttons = 32; 
```

Comment:
Do all these controllers actually have 32 buttons? This should be set to the actual number of buttons that can be pressed by the user.